woocommerce/woocommerce
# Include external feature plugins as packages

Adds support for external packages that hold core features. Currently this covers:

- [WooCommerce REST API package](https://github.com/woocommerce/woocommerce-rest-api)
- [Blocks](https://github.com/woocommerce/woocommerce-gutenberg-products-block)

The included `composer.json` file declares these and the versions used:

```json
  "require": {
    "composer/installers": "1.6.0",
    "woocommerce/woocommerce-rest-api": "dev-test/jetpack-autoloader",
    "woocommerce/woocommerce-blocks": "dev-build/2.2.0-dev",
    "automattic/jetpack-autoloader": "1.2.0"
  },
```

In the future, this will also be how [WooCommerce Admin](https://github.com/woocommerce/woocommerce-admin) is merged. 

Feature plugins can follow the [Example Package](https://github.com/woocommerce/woocommerce-example-package) template to be included in core as a package, as well as support standalone install as a plugin.

Packages are not included in version control, so developers using GitHub need to perform a build step after checking out:

```sh
$ composer install
```

This will pull in packages based on the versions defined in `composer.json`. Instructions have been added to [this doc](https://github.com/woocommerce/woocommerce/wiki/How-to-set-up-WooCommerce-development-environment) to cover Composer usage.

This PR needs to be updated with stable package versions (when they are tagged) but this is now working. Tests are passing. Note: REST API tests have been moved to the package. Only the package loader/integration is tested in core.

To test:
- Checkout this branch
- Run `composer install`
- Verify the `packages` directory is populated with the rest-api and blocks packages
- Verify functionality of WooCommerce works; you can also check the system status report which includes the rest api version and path

You can also install the rest-api-package as a plugin and verify it does not conflict. 

**Pull Requests to review in order:**
- [x] #23958 - Moving the legacy API into a separate directory away from the main REST API
- [x] https://github.com/woocommerce/woocommerce/pull/23959 - Remove core REST API classes
- [x] https://github.com/woocommerce/woocommerce/pull/23977 - Removes existing Gutenberg blocks code
- [x] https://github.com/woocommerce/woocommerce/pull/23978 - Add packages and the new autoloader
- [x] https://github.com/woocommerce/woocommerce/pull/23979 - Update direct usage of REST API controllers
- [x] https://github.com/woocommerce/woocommerce/pull/23981 - Init blocks package
- [x] https://github.com/woocommerce/woocommerce/pull/23989 - Update text domains for packages after composer update
- [x] https://github.com/woocommerce/woocommerce/pull/24002 - Unit tests

After review and merge of the above, the following items will remain:

- [x] Implement version and path retrieval in rest api package
- [x] Do text domains need find/replace?
- [x] (Possibly) remove all existing blocks/API tests from the unit test suite.
- [x] Add a unit test to check if built files exist
- [x] Run `composer install` before running tests in travis config.
- [x] Create a wiki doc explaining how to build (running composer) and update the build notice to link to this doc.
- [x] Update readme/contributing doc on composer and autoloader
- [x] Check core enforces PHP 5.6
- [x] Update blocks release docs
- [ ] Update blocks and rest-api versions to stable

There is also room to expand the autoloader to the entire 'includes' directory so core can remove the need to manually include everything. That's not something I'll be doing for this PR, but it's possible because I got it working locally :) cc @kloon @peterfabian
